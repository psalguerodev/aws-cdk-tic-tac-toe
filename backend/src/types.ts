export interface GameResult {
  playerX: PlayerStats;
  playerO: PlayerStats;
  winner: string | null;
  date: string;
  vsComputer: boolean;
  difficulty?: string;
}

export interface PlayerStats {
  name: string;
  wins: number;
}

export interface GameStats {
  games: GameResult[];
  totalGames: number;
  playerStats: {
    [key: string]: PlayerStats;
  };
}
