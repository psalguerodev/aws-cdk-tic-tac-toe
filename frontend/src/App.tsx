import { ChakraProvider, Container } from "@chakra-ui/react";
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import Home from "./pages/Home";
import Us from "./pages/Us";
import Links from "./pages/Links";

const queryClient = new QueryClient();

function App() {
  return (
    <QueryClientProvider client={queryClient}>
      <ChakraProvider>
        <Router>
          <Container maxW="100vw" p={0} centerContent>
            <Routes>
              <Route path="/" element={<Home />} />
              <Route path="/us" element={<Us />} />
              <Route path="/links" element={<Links />} />
            </Routes>
          </Container>
        </Router>
      </ChakraProvider>
    </QueryClientProvider>
  );
}

export default App;
