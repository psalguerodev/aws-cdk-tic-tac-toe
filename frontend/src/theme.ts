import { theme as baseTheme } from "@chakra-ui/react";

const theme = {
  ...baseTheme,
  styles: {
    global: {
      body: {
        bg: "gray.50",
      },
    },
  },
  components: {
    Container: {
      baseStyle: {
        maxW: "container.xl",
        px: [4, 6],
      },
    },
  },
};

export default theme;
