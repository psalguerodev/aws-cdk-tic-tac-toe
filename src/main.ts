import { App, Stack, StackProps, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { WebsiteConstruct } from './constructs/website-construct';

export class MyStack extends Stack {
  constructor(scope: Construct, id: string, props: StackProps = {}) {
    super(scope, id, props);

    // Crear el sitio web con S3 y CloudFront
    const website = new WebsiteConstruct(this, 'WebsiteConstruct', {
      bucketName: `${id}-website-bucket`,
      websitePath: '../../frontend/dist',
    });

    // Outputs a nivel de stack
    new CfnOutput(this, 'WebsiteURLStack', {
      value: website.websiteUrl,
      description: 'URL del sitio web (Stack)',
    });

    new CfnOutput(this, 'BucketNameStack', {
      value: website.bucketName,
      description: 'Nombre del bucket S3 (Stack)',
    });
  }
}

// for development, use account/region from cdk cli
const devEnv = {
  account: process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEFAULT_REGION,
};

const app = new App();

new MyStack(app, 'cdk-projen-io-ts-dev', { env: devEnv });
// new MyStack(app, 'cdk-projen-io-ts-prod', { env: prodEnv });

app.synth();
