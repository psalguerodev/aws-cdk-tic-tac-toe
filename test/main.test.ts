import { App } from 'aws-cdk-lib';
import { Template, Match } from 'aws-cdk-lib/assertions';
import { MyStack } from '../src/main';

jest.mock('aws-cdk-lib/aws-s3-deployment', () => ({
  BucketDeployment: jest.fn().mockImplementation(() => ({})),
  Source: {
    asset: jest.fn().mockImplementation(() => ({})),
  },
}));

describe('MyStack', () => {
  let app: App;
  let stack: MyStack;
  let template: Template;

  beforeEach(() => {
    app = new App();
    stack = new MyStack(app, 'test');
    template = Template.fromStack(stack);
  });

  test('crea el bucket S3 para el sitio web con la configuración correcta', () => {
    template.resourceCountIs('AWS::S3::Bucket', 1);
    template.hasResourceProperties('AWS::S3::Bucket', {
      WebsiteConfiguration: Match.objectLike({
        IndexDocument: 'index.html',
        ErrorDocument: 'index.html',
      }),
      PublicAccessBlockConfiguration: Match.objectLike({
        BlockPublicAcls: true,
        BlockPublicPolicy: true,
        IgnorePublicAcls: true,
        RestrictPublicBuckets: true,
      }),
    });
  });

  test('crea los outputs del stack', () => {
    template.hasOutput('WebsiteURLStack', {
      Description: 'URL del sitio web (Stack)',
    });

    template.hasOutput('BucketNameStack', {
      Description: 'Nombre del bucket S3 (Stack)',
    });
  });

  test('el bucket tiene la configuración CORS correcta', () => {
    template.hasResourceProperties('AWS::S3::Bucket', {
      CorsConfiguration: Match.objectLike({
        CorsRules: [
          {
            AllowedHeaders: ['*'],
            AllowedMethods: ['GET'],
            AllowedOrigins: ['*'],
          },
        ],
      }),
    });
  });

  test('snapshot del stack', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});
