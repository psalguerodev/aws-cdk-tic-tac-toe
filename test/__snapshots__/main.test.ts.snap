// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyStack snapshot del stack 1`] = `
{
  "Outputs": {
    "BucketNameStack": {
      "Description": "Nombre del bucket S3 (Stack)",
      "Value": {
        "Ref": "WebsiteConstructtestwebsitebucket05EDC685",
      },
    },
    "WebsiteConstructBucketName38C8193D": {
      "Description": "Nombre del bucket S3",
      "Value": {
        "Ref": "WebsiteConstructtestwebsitebucket05EDC685",
      },
    },
    "WebsiteConstructCloudFrontURLDC3C8449": {
      "Description": "URL de CloudFront",
      "Value": {
        "Fn::GetAtt": [
          "WebsiteConstructDistribution565587D6",
          "DomainName",
        ],
      },
    },
    "WebsiteURLStack": {
      "Description": "URL del sitio web (Stack)",
      "Value": {
        "Fn::GetAtt": [
          "WebsiteConstructtestwebsitebucket05EDC685",
          "WebsiteURL",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "WebsiteConstructDistribution565587D6": {
      "Properties": {
        "DistributionConfig": {
          "CustomErrorResponses": [
            {
              "ErrorCode": 403,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
            {
              "ErrorCode": 404,
              "ResponseCode": 200,
              "ResponsePagePath": "/index.html",
            },
          ],
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "testWebsiteConstructDistributionOrigin157C8D351",
            "ViewerProtocolPolicy": "redirect-to-https",
          },
          "DefaultRootObject": "index.html",
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Origins": [
            {
              "DomainName": {
                "Fn::GetAtt": [
                  "WebsiteConstructtestwebsitebucket05EDC685",
                  "RegionalDomainName",
                ],
              },
              "Id": "testWebsiteConstructDistributionOrigin157C8D351",
              "OriginAccessControlId": {
                "Fn::GetAtt": [
                  "WebsiteConstructDistributionOrigin1S3OriginAccessControl59E76E0A",
                  "Id",
                ],
              },
              "S3OriginConfig": {
                "OriginAccessIdentity": "",
              },
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "WebsiteConstructDistributionOrigin1S3OriginAccessControl59E76E0A": {
      "Properties": {
        "OriginAccessControlConfig": {
          "Name": "testWebsiteConstructDistribuOrigin1S3OriginAccessControl472B02BE",
          "OriginAccessControlOriginType": "s3",
          "SigningBehavior": "always",
          "SigningProtocol": "sigv4",
        },
      },
      "Type": "AWS::CloudFront::OriginAccessControl",
    },
    "WebsiteConstructtestwebsitebucket05EDC685": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "BucketName": "test-website-bucket",
        "CorsConfiguration": {
          "CorsRules": [
            {
              "AllowedHeaders": [
                "*",
              ],
              "AllowedMethods": [
                "GET",
              ],
              "AllowedOrigins": [
                "*",
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
        "WebsiteConfiguration": {
          "ErrorDocument": "index.html",
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "WebsiteConstructtestwebsitebucketPolicy36C312AB": {
      "Properties": {
        "Bucket": {
          "Ref": "WebsiteConstructtestwebsitebucket05EDC685",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "s3:GetObject",
              "Condition": {
                "StringEquals": {
                  "AWS:SourceArn": {
                    "Fn::Join": [
                      "",
                      [
                        "arn:",
                        {
                          "Ref": "AWS::Partition",
                        },
                        ":cloudfront::",
                        {
                          "Ref": "AWS::AccountId",
                        },
                        ":distribution/",
                        {
                          "Ref": "WebsiteConstructDistribution565587D6",
                        },
                      ],
                    ],
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Service": "cloudfront.amazonaws.com",
              },
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "WebsiteConstructtestwebsitebucket05EDC685",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
